lib/cosmos/gui/text/ruby_editor.rb

```diff
@@ -76,67 +76,63 @@ def self.create_format(color = Cosmos::BLACK, style='')
           # Ruby global
-          ['\\$\b\w+', 0, STYLES['string']],
+          ['\\$\b\w+', STYLES['string']],
           # Regex, possibly containing escape sequences
-          ["/.*/", 0, STYLES['string']],
+          ["/.*/", STYLES['string']],
           # Double-quoted string, possibly containing escape sequences
-          ['"([^"\\\\]|\\\\.)*"', 0, STYLES['string']],
+          ['"(?:[^"\\\\]|\\\\.)*"', STYLES['string']],
           # Match interpolated strings "blah #{code} blah"
-          ['(\#\\{)[^\\}]*\\}', 1, STYLES['brace']],
-          ['\#\\{[^\\}]*(\\})', 1, STYLES['brace']],
-          ['\#\\{([^\\}]*)\\}', 1, STYLES['normal']],
+          # This comes before single-quote and back-tick because
+          # those strings should not interpolate variables
+          ['(\#\\{)[^\\}]*\\}', STYLES['brace']],
+          ['\#\\{[^\\}]*(\\})', STYLES['brace']],
+          ['\#\\{([^\\}]*)\\}', STYLES['normal']],
           # Single-quoted string, possibly containing escape sequences
-          ["'[^'\\\\]*(\\.[^'\\\\]*)*'", 0, STYLES['string']],
+          ["'[^'\\\\]*(?:\\.[^'\\\\]*)*'", STYLES['string']],
           # Back-tick string, possibly containing escape sequences
-          ["`[^`\\\\]*(\\.[^`\\\\]*)*`", 0, STYLES['string']],
-          # A single # possibly followed by matched quotes
-          ['#(?=([^\'"]*(\'|")[^\'"]*(\'|"))*[^\'"]*$).*', 0, STYLES['comment']],
+          ["`[^`\\\\]*(\\.[^`\\\\]*)*`", STYLES['string']],
+          # Double quoted string | single quoted string | comment
+          # We don't want the first two but we want the last
+          ["\"[^\"]*\"|\'[^\']*\'|(#.*)", STYLES['comment']],
       ])
 
-      # Build a QRegExp for each pattern
+      # Build a RegExp for each pattern
       RULES_INFO = []
-      RULES.each do |pat, index, fmt|
-        RULES_INFO << [Qt::RegExp.new(pat, Qt::CaseSensitive, Qt::RegExp::RegExp2), index, fmt]
+      RULES.each do |pat, fmt|
+        RULES_INFO << [Regexp.new(pat), fmt]
       end
 
       def highlightBlock(text)
         # Do other syntax formatting
-        RULES_INFO.each do |expression, nth, format|
-          index = expression.indexIn(text, 0)
-          last_index = index
-          last_length = -1
-          while index >= 0
-            # We actually want the index of the nth match
-            index = expression.pos(nth)
-            # Bail if the index goes negative because it means
-            # we won't have a valid capture
-            break if index < 0
-            length = expression.cap(nth).length()
-            break if length <= 0
-            # Break if we're in an endless loop
-            break if (last_index == index and last_length == length)
-            last_length = length
-            last_index = index
-            self.setFormat(index, length, format)
-            index = expression.indexIn(text, index + length)
+        RULES_INFO.each do |expression, format|
+          match = text.scan(expression).flatten.compact
```

Comment:
Flatten compact smashes the (sometimes) multi-dimentional arrays and removes nils. It allowed me to eliminate the 0, 1 indexes from above. It also forced me to add non-capturing group syntax to a few of the matches but it makes the processing way easier.